src/typecheck.rs

```diff
@@ -1577,6 +1597,113 @@ fn constraint(state: &mut State, x: TypeWrapper, id: Ident) -> Result<(), RowUni
     }
 }
 
+/// Add row constraints on a freshly instantiated type variable.
+///
+/// When instantiating a quantified type variable with a unification variable, row constraints may
+/// apply. For example, if we instantiate `forall a. {x: Num | a} -> Num` by replacing `a` with a
+/// unification variable `Ptr(p)`, this unification variable requires a constraint to avoid being
+/// unified with a row type containing another declaration for the field `x`.
+///
+/// # Preconditions
+///
+/// Because `constraint_var` should be called on a fresh unification variable `p`, the following is
+/// assumed:
+/// - `get_root(state.table, p) == p`
+/// - `state.constr.get(&p) == None`
+fn constrain_var(state: &mut State, tyw: &TypeWrapper, p: usize) {
+    fn constrain_var_(state: &mut State, mut constr: HashSet<Ident>, tyw: &TypeWrapper, p: usize) {
```

Comment:
I'm not fully sure what is happening here. But I think that what you are doing is, when instantiating a `forall a. t` with a unification variable, you traverse `t` in search of row constraints. I fear that this may be a bit costly (there is some quadraticness to it).

An alternative strategy is to simply generate a unification variable, which doesn't have constraints. Then, when unifying two types, one of them being a record type, you can
1. generate a new type variable for this row, with the appropriate row constraints
2. unify the row variable with the new variable

Reply:
>I'm not fully sure what is happening here. But I think that what you are doing is, when instantiating a forall a. t with a unification variable, you traverse t in search of row constraints. I fear that this may be a bit costly (there is some quadraticness to it).

Yes, indeed.  For a type `forall a1 ... an . type`, it's `O(n*size(type))`.

If I understand correctly your suggestion, you propose to do this (constraint generation) at unification time, rather than instantiation time, so that:
1. Non-row parts of the type are not uselessly visited. Currently (this PR), instantiating `forall a b c d. Num -> Num  -> Num -> a -> b -> c -> d` does cause some useless work.
2. Rows that are not unified are not treated, while they currently are.

On the other hand, would that be possible that it duplicates some work in other instances ? First guess is in a situation where a unification would be immediate otherwise (say, unify a unification variable with a row type) and repeated on a same row type (several consecutive unification of different variables with the same row type). Then the row type would be visited many times in vain (the first visit is the. Not sure it can happen though (other than cheating with `Assume` to create otherwise uninhabited polymorphic types, like `Assume(forall a. {field: Num | a}, "I'm lying")`).

Another solution is to do type variable substitution and constraint generation in one pass, as substitution already traverses the type anyway.

Reply:
Ah yes… substitution. In an ideal world, substitution would be performed lazily. If we can't have this, we would like to have a single substitution for all the variables. But, yes, I suppose that's an unavoidable cost, so maybe it works just as well to infer row constraints during substitution.